internal/download/get.go

```diff
@@ -31,81 +34,115 @@ func init() {
 // SetMocking sets the correct Get methods for testing
 func SetMocking(useMocking bool) {
 	if useMocking {
-		Get = _testHTTPGet
-		GetWithProgress = _testHTTPGetWithProgress
+		GetWithProgress = _testGetWithProgress
 	} else {
-		Get = httpGet
-		GetWithProgress = httpGetWithProgress
+		GetWithProgress = s3GetWithProgress
 	}
 }
 
-func httpGet(url string) ([]byte, *failures.Failure) {
-	logging.Debug("Retrieving url: %s", url)
-	return httpGetWithProgress(url, nil)
-}
+func s3GetWithProgress(url *url.URL, progress *progress.Progress) ([]byte, error) {
+	logging.Debug("Downloading via s3")
 
-func httpGetWithProgress(url string, progress *progress.Progress) ([]byte, *failures.Failure) {
-	logging.Debug("Retrieving url: %s", url)
-	client := retryhttp.NewClient(0 /* 0 = no timeout */, 5)
-	resp, err := client.Get(url)
-	if err != nil {
-		code := -1
-		if resp != nil {
-			code = resp.StatusCode
-		}
-		return nil, failures.FailNetwork.Wrap(err, locale.Tl("err_network_get", "Status code: {{.V0}}", strconv.Itoa(code)))
-	}
-	defer resp.Body.Close()
+	s3m := parseS3URL(url)
 
-	if resp.StatusCode != 200 {
-		return nil, failures.FailNetwork.New("err_invalid_status_code", strconv.Itoa(resp.StatusCode))
+	// Prepare AWS session
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String(s3m.Region),
+		Credentials: credentials.AnonymousCredentials,
+	})
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not create aws session")
 	}
 
-	var total int
-	length := resp.Header.Get("Content-Length")
-	if length == "" {
-		total = 1
+	// Grab file size
+	var length int64 = 0
+	res, err := http.Get(url.String())
+	if err != nil {
+		logging.Debug("Could not grab url: %v", err)
+	} else if res.StatusCode != http.StatusOK {
+		logging.Debug("Could not grab url due to statuscode: %d", res.StatusCode)
 	} else {
-		total, err = strconv.Atoi(length)
+		lengthInt, err := strconv.Atoi(res.Header.Get("Content-Length"))
 		if err != nil {
-			logging.Debug("Content-length: %v", length)
-			return nil, failures.FailInput.Wrap(err)
+			logging.Debug("Could not grab content-length: %v", err)
+		} else {
+			length = int64(lengthInt)
 		}
 	}
```

Comment:
Is it OK that the length is potentially 0? I think a basic default value (128?) would at least give some indication of action despite being wrong. Otherwise, 0 will show the bar at 100% and will be wrong, but it will show no action at all.

Reply:
Yeah, I tested with 0. It'll still show you it's pulling in stuff but you won't know how close it is to finishing, which in this scenario we can't know.